proposals/collection-literals.md

```diff
@@ -0,0 +1,1354 @@
+        "-cp", dependencyFiles.joinToString(File.pathSeparator),
+        "-module-name", moduleName,
+    ]
+
+
+// before 3
+    projectTest(
+        parallel = true,
+        defineJDKEnvVariables = listOf(JdkMajorVersion.JDK_1_8, JdkMajorVersion.JDK_11_0, JdkMajorVersion.JDK_17_0)
+    ) { /* ... */ }
+// after 3
+    projectTest(
+        parallel = true,
+        defineJDKEnvVariables = [JdkMajorVersion.JDK_1_8, JdkMajorVersion.JDK_11_0, JdkMajorVersion.JDK_17_0]
+    ) { /* ... */ }
+
+
+// before 4
+    val toReplace = mutableListOf<String>()
+    val visited = mutableSetOf<String>()
+// after 4
+    val toReplace: MutableList<String> = []
+    val visited: MutableSet<String> = []
+
+
+// before 5
+    it[AnalysisFlags.optIn] = optInList + listOf("kotlin.ExperimentalUnsignedTypes")
+// after 5
+    it[AnalysisFlags.optIn] = optInList + ["kotlin.ExperimentalUnsignedTypes"]
+
+
+// before 6
+    for (key in listOf(argument.value, argument.shortName, argument.deprecatedName)) {
+        if (key.isNotEmpty()) put(key, argumentField)
+    }
+// after 6
+    for (key in [argument.value, argument.shortName, argument.deprecatedName]) {
+        if (key.isNotEmpty()) put(key, argumentField)
+    }
+
+
+// before 7
+    override fun getMimeTypes(): List<String> = listOf("text/x-kotlin")
+// after 7
+    override fun getMimeTypes(): List<String> = ["text/x-kotlin"]
+
+
+// before 8
+    fun <D> MutableMap<String, MutableSet<D>>.initAndAdd(key: String, value: D) {
+        this.compute(key) { _, maybeValues -> (maybeValues ?: mutableSetOf()).apply { add(value) } }
+    }
+// after 8
+    fun <D> MutableMap<String, MutableSet<D>>.initAndAdd(key: String, value: D) {
+        this.compute(key) { _, maybeValues -> (maybeValues ?: []).apply { add(value) } }
+    }
+
+
+// before 9
+    modules.mapNotNullTo(hashSetOf()) { environment.findLocalFile(it.getOutputDirectory()) }
+    sourcesByModuleName.getOrPut(moduleName) { mutableSetOf() }.add(sourceFile)
+// after 9
+    // Can't be expressed via collection literals
+
+
+// before 10
+    if (asterisk[row] != emptyList<Int>()) { /* ... */ }
+// after 10
+    // Can't be expressed via collection literals
+```
+
+## Proposal
+
+**Informally**, the proposal strives to make it possible for users to use collection literals syntax to express user-defined types `val foo: MyCustomList<Int> = [1, 2, 3]`.
+And when the *expected type* (See the definition below) is unspecified, expression must fall back to the `kotlin.List` type: `val foo = [1, 2, 3] // List<Int>`.
+
+It's proposed to use square brackets because the syntax is already used in Kotlin for array literals inside annotation constructor arguments.
+It's the syntax a lot of programmers are already familiar with coming from other programming languages,
+and because it honors mathematical notation for matrices.
+
+```kotlin
+class MyCustomList<T> {
+    companion object { fun <T> of(vararg elements: T): MyCustomList<T> = TODO() }
```

Comment:
Based on the examples further down the proposal, I believe `operator` should be used here too.

```suggestion
    companion object { operator fun <T> of(vararg elements: T): MyCustomList<T> = TODO() }
```

Reply:
Thanks, fixed